File: lib/ts/scooter-schema/src/guard.ts
/**
 * Runtime ownership guard. Each service connects to exactly one shared database
 * (the one it owns / co-owns); this asserts a client is actually pointed at that
 * database, so a mis-wired DSN fails fast at startup instead of writing into the
 * wrong database's tables. Pair it with the per-database table exports from this
 * package — a service should only import the database module(s) it may touch.
 *
 * NOT generated — hand-written and stable across `just db-generate`.
 */

/** The database names this package models — the atlas envs under lib/sql. */
export const DATABASES = ["webhooks", "scheduler", "broker", "byoc", "agent_host"] as const;
export type Database = (typeof DATABASES)[number];

/** Minimal shape of a node-postgres Pool/Client — just enough to run one query. */
export interface Queryable {
  query(text: string): Promise<{ rows: Array<Record<string, unknown>> }>;
}

/**
 * Assert `db` is connected to the `expected` database. Throws otherwise. Call it
 * once at startup (best-effort is fine: a wrong database is a deploy misconfig,
 * not a transient error, so failing loudly is correct).
 */
export async function assertDatabase(db: Queryable, expected: Database): Promise<void> {
  const res = await db.query("select current_database() as db");
  const actual = res.rows[0]?.db;
  if (actual !== expected) {
    throw new Error(
      `@scooter/schema: connected to database "${String(actual)}" but this client is for "${expected}" — refusing to run. A service must only use the database it owns (check its DSN's DB name).`,
    );
  }
}

File: lib/py/scooter-schema/src/scooter_schema/guard.py
"""Runtime ownership guard: assert a connection is pointed at the database this
service owns, so a mis-wired DSN fails fast at startup instead of writing into
another database's tables. NOT generated.
"""

from __future__ import annotations

DATABASES = ("webhooks", "scheduler", "broker", "byoc")


def check_database(actual: str, expected: str) -> None:
    """Raise if ``actual`` (the connected database) is not ``expected``. Pure —
    the testable core the async/sync helpers below wrap."""
    if actual != expected:
        raise RuntimeError(
            f'scooter_schema: connected to database "{actual}" but this service owns '
            f'"{expected}" — refusing to run. Check the DSN\'s database name.'
        )


async def assert_database(conn, expected: str) -> None:
    """Assert an async SQLAlchemy connection is on ``expected``. ``conn`` is an
    ``AsyncConnection`` (e.g. ``async with engine.connect() as conn``)."""
    from sqlalchemy import text

    result = await conn.execute(text("select current_database()"))
    check_database(result.scalar_one(), expected)


def assert_database_sync(conn, expected: str) -> None:
    """Assert a sync SQLAlchemy connection is on ``expected``."""
    from sqlalchemy import text

    result = conn.execute(text("select current_database()"))
    check_database(result.scalar_one(), expected)
